### DISCUSSION TITLE
[Help] Toggle flag binding across live grep modes

### DISCUSSION THREAD

**Discussion by Ajaymamtora** [ibhagwan/fzf-lua#1488] [Q&A] [answered]
I found this snippet which allows us to toggle a flag

```lua
M.rg_toggle_flag = function(flag)
  return function(_, opts)
    local o = { resume = true, cwd = opts.cwd }
    if not flag:match("^%s") then
      -- flag must be preceded by whitespace
      flag = " " .. flag
    end
    -- grep|live_grep sets `opts._cmd` to the original
    -- command without the search argument
    local cmd = opts._cmd or opts.cmd
    if cmd:match(require("fzf-lua.utils").lua_regex_escape(flag)) then
      o.cmd = cmd:gsub(require("fzf-lua.utils").lua_regex_escape(flag), "")
    else
      local bin, args = cmd:match("([^%s]+)(.*)$")
      o.cmd = string.format("%s%s%s", bin, flag, args)
    end
    opts.__call_fn(o)
  end
end
```

But when using it in lgrep_curbuf, it resets the scope to cwd, not the current file.

Is there a proper way to toggle a flag when live_grepping that retains all the other current settings like flags + search paths etc?

**Comment by ibhagwan** [comment]
You can customize which flag to toggle by setting `grep.toggle_ignore_flag` and using the `toggle_ignore` action or set `toggle_hiden_flag` and use the `toggle_hidden` action.

**Reply by Ajaymamtora** [reply]
Does this work for toggling 2 case sensitivity flags ?

I want to alternate between --ignore-case and --case-sensitive as well for example

**Reply by ibhagwan** [reply]
It might, try.

If it doesn’t work copy the code from actions the make a custom toggle:
https://github.com/ibhagwan/fzf-lua/blob/1e03541de4d8a169defe83bb4d7abfba450c63a1/lua/fzf-lua/actions.lua#L825-L846

**Reply by Ajaymamtora** [reply]
I've tried both of these and searching through github and can't figure out how to build a working flag toggling function.

Here's what i've tried so far:

```lua
M.rg_toggle_flag = function(flag)
  local utils = require("fzf-lua.utils")
  return function(_, opts)
    local o = { resume = true, cwd = opts.cwd }
    if not flag then
      utils.err("'toggle_flag' not set")
      return
    end
    if not flag:match("^%s") then
      -- flag must be preceded by whitespace
      flag = " " .. flag
    end
    -- grep|live_grep sets `opts._cmd` to the original
    -- command without the search argument
    local cmd = opts._cmd or opts.cmd
    if cmd:match(utils.lua_regex_escape(flag)) then
      o.cmd = cmd:gsub(utils.lua_regex_escape(flag), "")
    else
      local bin, args = cmd:match("([^%s]+)(.*)$")
      o.cmd = string.format("%s%s%s", bin, flag, args)
    end
    opts.__call_fn(o)
  end
end

or 

M.rg_toggle_flag = function(flag)
  return function(_, opts)
    local o = { resume = true, cwd = opts.cwd }
    if not flag:match("^%s") then
      -- flag must be preceded by whitespace
      flag = " " .. flag
    end
    -- grep|live_grep sets `opts._cmd` to the original
    -- command without the search argument
    local cmd = opts._cmd or opts.cmd
    if cmd:match(require("fzf-lua.utils").lua_regex_escape(flag)) then
      o.cmd = cmd:gsub(require("fzf-lua.utils").lua_regex_escape(flag), "")
    else
      local bin, args = cmd:match("([^%s]+)(.*)$")
      o.cmd = string.format("%s%s%s", bin, flag, args)
    end
    opts.__call_fn(o)
  end
end

-- Wrapper function for toggle_flag action
M.toggle_flag = function(flag)
  return function(_, opts)
    local actions = require("fzf-lua.actions")

    -- Create a new options table with the toggle_flag
    local new_opts = vim.tbl_extend("force", opts or {}, { toggle_flag = flag })

    -- Call the internal toggle_flag action
    actions.toggle_flag(_, new_opts)

    -- Resume the fzf-lua session to reflect the changes
    fzf_lua.resume()
  end
end
```

Please can you implement a flag toggle function the user can call from the outside that does not lose the fzf state? rg_toggle_flag is close but like i said, when using lgrep_curbuf, it resets the scope to cwd, not the current file.

**Reply by ibhagwan** [reply]
> Please can you implement a flag toggle function the user can call from the outside that does not lose the fzf state? rg_toggle_flag is close but like i said, when using lgrep_curbuf, it resets the scope to cwd, not the current file.

Sure, should be an easy bugfix to maintain the filename / cwd in the toggle although it might take me a couple of weeks to get to as I’m traveling.

**Reply by ibhagwan** [reply]
#1489 if you want to track this.

**Reply by Ajaymamtora** [reply]
> > Please can you implement a flag toggle function the user can call from the outside that does not lose the fzf state? rg_toggle_flag is close but like i said, when using lgrep_curbuf, it resets the scope to cwd, not the current file.
> 
> 
> 
> Sure, should be an easy bugfix to maintain the filename / cwd in the toggle although it might take me a couple of weeks to get to as I’m traveling.

Thank you! No worries

**Reply by ibhagwan** [reply]
@Ajaymamtora, first update to https://github.com/ibhagwan/fzf-lua/commit/fc0f2b5781587d388b743c2c114127a7036d1a31, then use:
```lua
require("fzf-lua").setup({
  grep = {
    actions = {
      ["alt-g"] = {
        fn = function(_, opts)
          require("fzf-lua").actions.toggle_flag(_, vim.tbl_extend("force", opts, {
            toggle_flag = "--ignore-case --case-sensitive"
          }))
        end,
        desc = "toggle-flags",
        header = "Toggle flags",
      },
    },
  },
})

```

**Reply by Ajaymamtora** [reply]
> @Ajaymamtora, first update to [fc0f2b5](https://github.com/ibhagwan/fzf-lua/commit/fc0f2b5781587d388b743c2c114127a7036d1a31), then use:
> 
> ```lua
> require("fzf-lua").setup({
>   grep = {
>     actions = {
>       ["alt-g"] = {
>         fn = function(_, opts)
>           require("fzf-lua").actions.toggle_flag(_, vim.tbl_extend("force", opts, {
>             toggle_flag = "--ignore-case --case-sensitive"
>           }))
>         end,
>         desc = "toggle-flags",
>         header = "Toggle flags",
>       },
>     },
>   },
> })
> ```

Thank you, i've tried it but doesnt seem to work.

These are my opts

```lua
    grep = {
      prompt = " ",
      rg_opts = "--column --line-number --no-heading --color=always --no-ignore --hidden --ignore-case --max-columns=4096 -e",
      no_header_i = true,
      rg_glob = true,
      exec_empty_query = true,
      actions = {
        ["alt-c"] = {
          fn = function(_, opts)
            require("fzf-lua").actions.toggle_flag(
              _,
              vim.tbl_extend("force", opts, {
                toggle_flag = "--ignore-case --case-sensitive",
              })
            )
          end,
          desc = "toggle-flags",
          header = "Toggle flags",
        },
      },
    },

```

which should remove ignore-case and apply case-sensitive right?

**Reply by ibhagwan** [reply]
> which should remove ignore-case and apply case-sensitive right?

Remove `--ignore-case` from your rg_opts, you’re setting as a permanent flag this way.

Can you test with `debug=true`? It should add the flags right after `rg`.

**Reply by Ajaymamtora** [reply]
> > which should remove ignore-case and apply case-sensitive right?
> 
> Remove `--ignore-case` from your rg_opts, you’re setting as a permanent flag this way.
> 
> Can you test with `debug=true`? It should add the flags right after `rg`.

Okay I tried that, called FzfLua lgrep_curbuf, searched 'REQ' then toggled, and it doesnt return any of the matches for lowercase 'REQ' (for require as an example.) Not sure if Im missing something

**Reply by ibhagwan** [reply]
> Okay I tried that, called FzfLua lgrep_curbuf, searched 'REQ' then toggled, and it doesnt return any of the matches for lowercase 'REQ' (for require as an example.) Not sure if Im missing something

Can you run with `:FzfLua lgrep_curbuf debug=true` and explore the command in the first line with `[DEBUG]` prefix? Post a screenshot of the command before and after the toggle.

**Reply by ibhagwan** [reply]
Why are you toggling `--case-sensitive` it’s the default, also need to remove smart case as it conflicts, use this:
```lua

require("fzf-lua").setup({
  grep = {
    rg_opts = "--column --line-number --no-heading --color=always --no-ignore --hidden --max-columns=4096 -e",
    actions = {
      ["alt-g"] = {
        fn = function(_, opts)
          require("fzf-lua").actions.toggle_flag(_, vim.tbl_extend("force", opts, {
            toggle_flag = "--ignore-case"
          }))
        end,
        desc = "toggle-flags",
        header = "Toggle flags",
      },
    },
  },
})
```

With the above setup `:FzfLua lgrep_curbuf debug=true fzf_opts.--with-nth=false`:

![IMG_2596](https://github.com/user-attachments/assets/5dda36fb-f1f5-4477-bd24-ff7351675239)

**Reply by Ajaymamtora** [reply]
> Why are you toggling `--case-sensitive` it’s the default, also need to remove smart case as it conflicts, use this:
> 
> ```lua
> require("fzf-lua").setup({
>   grep = {
>     rg_opts = "--column --line-number --no-heading --color=always --no-ignore --hidden --max-columns=4096 -e",
>     actions = {
>       ["alt-g"] = {
>         fn = function(_, opts)
>           require("fzf-lua").actions.toggle_flag(_, vim.tbl_extend("force", opts, {
>             toggle_flag = "--ignore-case"
>           }))
>         end,
>         desc = "toggle-flags",
>         header = "Toggle flags",
>       },
>     },
>   },
> })
> ```
> 
> With the above setup `:FzfLua lgrep_curbuf debug=true fzf_opts.--with-nth=false`:
> 
> ![IMG_2596](https://private-user-images.githubusercontent.com/59988195/379237530-5dda36fb-f1f5-4477-bd24-ff7351675239.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3MTgwMDMsIm5iZiI6MTcyOTcxNzcwMywicGF0aCI6Ii81OTk4ODE5NS8zNzkyMzc1MzAtNWRkYTM2ZmItZjFmNS00NDc3LWJkMjQtZmY3MzUxNjc1MjM5LmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyM1QyMTA4MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZTBjZTkxYTNhM2MyZmVhYjc4OTI5Y2JiOGU2YzJiZGU0MzRmYmI4YzZhYmI5YjgxNjEyN2E0M2IzNmM4N2FlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.F8MKdLcDXLgkfPQqkLs6KkO0UGgjfTTT9REUx93zMLs)

Are you saying --case-sensitive is applied by default even though ive specified rg_opts without it?

Also I copied your exact config and it doesnt toggle any flags in the debug line, 

Heres what I see

![image](https://github.com/user-attachments/assets/92acd609-07a5-4eb2-ab91-97a83b285269)

**Reply by ibhagwan** [reply]
> Are you saying --case-sensitive is applied by default even though ive specified rg_opts without it?

It’s not applied by fzf-lua, read man rg, default search is case sensitive.

> Also I copied your exact config and it doesnt toggle any flags in the debug line,

Did you press alt-g?

**Reply by Ajaymamtora** [reply]
> > Are you saying --case-sensitive is applied by default even though ive specified rg_opts without it?
> 
> It’s not applied by fzf-lua, read man rg, default search is case sensitive.
> 
> > Also I copied your exact config and it doesnt toggle any flags in the debug line,
> 
> Did you press alt-g?

Yes it reloaded the list as well. The debug command on the left didnt change at all when I toggled it. When typing, I could see it being added to debug command so I know it was updating

**Reply by ibhagwan** [reply]
> Yes it reloaded the list as well. The debug command on the left didnt change at all when I toggled it. When typing, I could see it being added to debug command so I know it was updating

I’m not sure why, the code is very simple, copy paste the code from `actions.toggle_flag` and use old fashioned print debugging to see why the flag isn’t added to the cmd.

**Comment by mmirus** [comment, marked as answer]
@Ajaymamtora in case you never found a good solution, here is what I just implemented based on @ibhagwan's suggestions above and the info in [this issue](https://github.com/ibhagwan/fzf-lua/pull/1502#issuecomment-2438384250).

It defaults to `--smart-case`, and you can toggle it with `alt-g`. The action help shown above the results will update to indicate whether `alt-g` will disable or enable smart case.

```lua
      grep = {
        rg_opts = "--column --line-number --no-heading --color=always --smart-case --max-columns=4096 --hidden -e",
        actions = {
          ["alt-g"] = {
            fn = function(_, opts)
              fzf.actions.toggle_flag(
                _,
                vim.tbl_extend("force", opts, {
                  toggle_flag = "--smart-case",
                })
              )
            end,
            desc = "toggle-flags",
            header = function(o)
              local flag = o.toggle_smart_case_flag or "--smart-case"
              if o.cmd and o.cmd:match(fzf.utils.lua_regex_escape(flag)) then
                return "Disable smart case"
              else
                return "Enable smart case"
              end
            end,
          },
        },
      },
```

Here it is in action:

https://github.com/user-attachments/assets/6ea15de5-98da-4df3-8ca5-9255bae96695

**Reply by Ajaymamtora** [reply]
Thanks that looks good, I've swapped back to telescope though
